manual/manual/refman/exten.etex

```diff
@@ -2461,4 +2461,119 @@ val x: int
   [@@@ocaml.alert deprecated "Please do something else"]
 \end{verbatim}
 
+\section{Let operators\label{s:let-operators} }
+(Introduced in 4.08.0)
+
+\begin{syntax}
+let-operator:
+ | 'let' ('$'||'&'||'*'||'+'||'-'||'/'||'<'||'='||'>'||'@'||'^'||'|') { operator-char }
+;
+and-operator:
+ | 'and' ('$'||'&'||'*'||'+'||'-'||'/'||'<'||'='||'>'||'@'||'^'||'|') { operator-char }
+;
+operator-name :
+          ...
+        | let-operator
+        | and-operator
+;
+expr:
+          ...
+        | let-operator let-binding { and-operator let-binding } in expr
+;
+\end{syntax}
+
+This extension provides a convenient syntax for working with monads and
+applicatives. Users can define {\em let operators}:
+
+\begin{caml_example}{verbatim}
+let ( let* ) o f =
+  match o with
+  | None -> None
+  | Some x -> f x
+
+let return x = Some x
+\end{caml_example}
+
+and then apply them using this convenient syntax:
+
+\begin{caml_example}{verbatim}
+let find_and_double tbl k =
+  let* x = Hashtbl.find_opt tbl k in
+    return (x * 2)
+\end{caml_example}
+
+which is equivalent to this expanded form:
+
+\begin{caml_example}{verbatim}
+let find_and_double tbl k =
+  ( let* ) (Hashtbl.find_opt tbl k)
+    (fun x -> return (x * 2))
+\end{caml_example}
+
+Users can also define {\em and operators}:
+
+\begin{caml_example}{verbatim}
+let ( and* ) o1 o2 =
+  match o1, o2 with
+  | None, _ -> None
+  | _, None -> None
+  | Some x1, Some x2 -> Some (x1, x2)
+\end{caml_example}
+
+to support the syntax:
+
+\begin{caml_example}{verbatim}
+let find_and_add tbl1 tbl2 k =
+  let* x1 = Hashtbl.find_opt tbl1 k
+  and* x2 = Hashtbl.find_opt tbl2 k in
+    return (x1 + x2)
+\end{caml_example}
+
+which is equivalent to this expanded form:
+
+\begin{caml_example}{verbatim}
+let find_and_add tbl1 tbl2 k =
+  let o1 = Hashtbl.find_opt tbl1 k
+  and o2 = Hashtbl.find_opt tbl2 k in
+  ( let* ) (( and* ) o1 o2)
+    (fun (x1, x2) -> return (x1 + x2))
+\end{caml_example}
+
+\begin{caml_eval}
+type 'a t
+\end{caml_eval}
+
+For monads, the intention is that "(let*)" is bound to the "bind"
+operation, and "(and*)" is bound to the monoidal product operation:
+
+\begin{caml_example*}{signature}
+val ( let* ) : 'a -> ('a -> 'b t) -> 'b t
+val ( and* ): 'a t -> 'b t -> ('a * 'b) t
+\end{caml_example*}
+
+For applicatives, the intention is that "(let+)" is bound to the "map"
+operation, and "(and+)" is again bound to the monoidal product
+operation:
+
+\begin{caml_example*}{signature}
+val ( let+ ) : 'a t -> ('a -> 'b) -> 'b t
+val ( and+ ): 'a t -> 'b t -> ('a * 'b) t
+\end{caml_example*}
+
+which allows the above example to be written as:
+
+\begin{caml_example}{verbatim}
+let ( let+ ) o f =
+  match o with
+  | None -> None
+  | Some x -> Some (f x)
+
+let ( and+ ) = ( and* )
+
+let find_and_add tbl1 tbl2 k =
+  let+ x1 = Hashtbl.find_opt tbl1 k
+  and+ x2 = Hashtbl.find_opt tbl2 k in
+    x1 + x2
+\end{caml_example}
```

Comment:
I'm still a bit unconvinced by the choice of examples here, both for the monad and the applicative. (Many people are going to read this documentation, so it is worth obsessing a bit about it.)

First, you should emphasize the fact that `let* p1 = e1 and* p2 = e2` will always compute both `e1` and `e2` (morally `and` computes them "in parallel" in some sense).

As a consequence, the use of `and*` in the option monad is fairly perplexing to me. The two following programs are not equivalent:

```
let* x1 = Hashtbl.find_opt tbl1 k
and* x2 = Hashtbl.find_opt tbl2 k in
return (x1 + x2);;

let* x1 = Hashtbl.find_opt tbl1 k in
let* x2 = Hashtbl.find_opt tbl2 k in
return (x1 + x2);;
```

and one should always use the second program instead of the first, as it performs strictly less work in case the `tbl1` lookup fails. This suggests that `option` is not the right example to present `and`-operators.

For the applicative example, could you maybe use a different functor than for the monad example? (Parser combinators?)

Reply:
I'm definitely not explaining parser combinators in the manual. Not sure there's a simple example using only simple types. I suppose `List` although I think that tends to confuse people as they are not used to thinking of a list as a computation.

Reply:
Maybe `Seq.t` would work? I know bind is expensive for some sequence types, although I'm not sure about the one in the stdlib.

Reply:
No. It looks like the applicative is basically the same as bind for `seq`.

Reply:
There is an `'a list` instance for applicative that zips elements pointwise, and is not a monad. Maybe you could use this (with another sequence structure than lists to avoid confusion), showing how to build a `map5` this way?

Reply:
Ok. What about if I did a version of Haskell's ZipList using `Seq.t`? That's something fairly simple. Although since it isn't a monad I can't use it to show `let*...and*` only `let+...and+`.

Reply:
Oh, now I see you had the same idea. I'll give that a go tomorrow then.

Reply:
Ok, I've updated the examples.

Reply:
@gasche Does the new example address your concerns?

Reply:
Yes, the new example is much nicer, thank you.